src/components/AutoSuggest/index.js

```diff
@@ -0,0 +1,252 @@
+  const [activeIndex, setActiveIndex] = useState(-1);
+  const wrapperRef = useRef(null);
+  const inputRef = useRef(null);
+
+  const handleKeyDown = useCallback(
+    event => {
+      const numberOfOptions = numOptionsDeterminer(data);
+      if (!show) return;
+
+      switch (event.keyCode) {
+        // Arrow up
+        case 38:
+          event.preventDefault();
+
+          setActiveIndex(state => {
+            const indexOfActive = state - 1;
+            const topReached = indexOfActive === -1;
+
+            return topReached ? numberOfOptions - 1 : indexOfActive;
+          });
+          break;
+
+        // Arrow down
+        case 40:
+          event.preventDefault();
+
+          setActiveIndex(state => {
+            const indexOfActive = state + 1;
+            const endReached = indexOfActive === numberOfOptions;
+
+            return endReached ? 0 : indexOfActive;
+          });
+          break;
+
+        // Enter
+        case 13:
+          if (event.target !== inputRef.current) {
+            inputRef.current.value = event.target.innerText;
+          }
+
+          setActiveIndex(-1);
+          setShow(false);
+          break;
+
+        // Escape
+        case 27:
+          inputRef.current.value = '';
+          setActiveIndex(-1);
+          setShow(false);
+          break;
+
+        // Home
+        // Moves focus to the textbox and places the editing cursor at the beginning of the field.
+        case 36:
+          event.preventDefault();
+
+          inputRef.current.focus();
+          inputRef.current.setSelectionRange(0, 0);
+          break;
+
+        // End
+        // Moves focus to the textbox and places the editing cursor at the end of the field.
+        case 35:
+          event.preventDefault();
+
+          inputRef.current.focus();
+          inputRef.current.setSelectionRange(9999, 9999);
+          break;
+
+        default:
+          setActiveIndex(-1);
+          break;
+      }
+    },
+    [data, numOptionsDeterminer, show]
+  );
+
+  useEffect(() => {
+    if (activeIndex === -1) {
+      inputRef.current.focus();
+    }
+  }, [activeIndex]);
+
+  useEffect(() => {
+    const wrapper = wrapperRef.current;
+
+    wrapper.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      wrapper.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [handleKeyDown]);
+
+  useEffect(() => {
+    const hasResults = numOptionsDeterminer(data) > 0;
+
+    setShow(hasResults);
+  }, [data, numOptionsDeterminer]);
+
+  const delayedCallback = useCallback(
+    debounce(value => {
+      serviceRequest(value);
+    }, INPUT_DELAY),
+    [serviceRequest]
+  );
+
+  const onChange = useCallback(
+    event => {
+      event.persist();
+      const { value } = event.target;
+      delayedCallback(value);
+    },
+    [delayedCallback]
+  );
+
+  const serviceRequest = useCallback(
+    value => {
+      if (value.length >= 3) {
+        get(`${url}${value}`);
```

Comment:
You are assuming here that the call format of the search service will always match this pattern. Maybe passing an url formatter instead of `url` as a function parameter would give us more flexibility.

Reply:
Good suggestion. However, I think that this format will cover all cases that we need for now. If we need a more intricate way of constructing the URL, we can always change it later on.